ksp_plugin/plugin.cpp

```diff
@@ -149,127 +124,120 @@ Plugin::~Plugin() {
-                       << "of initialization";
-  CHECK(!absolute_initialization_);
-  CHECK_EQ((bool)parent_index, (bool)keplerian_elements);
-  CHECK_EQ((bool)parent_index, (bool)hierarchical_initialization_);
-  RotatingBody<Barycentric>* const unowned_body = body.get();
-  if (hierarchical_initialization_) {
-    hierarchical_initialization_->system.Add(
-        std::move(body),
-        hierarchical_initialization_->indices_to_bodies[*parent_index],
-        *keplerian_elements);
-  } else {
-    hierarchical_initialization_.emplace(std::move(body));
+    serialization::GravityModel::Body const& gravity_model,
+    serialization::InitialState::Keplerian::Body const& initial_state) {
+  CHECK_EQ(gravity_model.name(), initial_state.name());
+  InitializeIndices(gravity_model.name(), celestial_index, parent_index);
+  *gravity_model_.add_body() = gravity_model;
+  CHECK(!initial_state_.has_cartesian()) << initial_state_.DebugString();
+  serialization::InitialState::Keplerian::Body* const body =
+      initial_state_.mutable_keplerian()->add_body();
+  *body = initial_state;
+  if (parent_index) {
+    body->set_parent(FindOrDie(index_to_name_, *parent_index));
   }
-  bool inserted =
-      hierarchical_initialization_->parents.emplace(celestial_index,
-                                                    parent_index).second;
-  inserted &=
-      hierarchical_initialization_->
-          indices_to_bodies.emplace(celestial_index, unowned_body).second;
-  CHECK(inserted);
-
-  // Record the fingerprints of the parameters to detect if we are in KSP stock.
-  CHECK(celestial_jacobi_keplerian_fingerprints_.insert(
-            FingerprintCelestialJacobiKeplerian(celestial_index,
-                                                parent_index,
-                                                keplerian_elements,
-                                                *unowned_body)).second);
 }
 
 void Plugin::EndInitialization() {
   CHECK(initializing_);
-  if (hierarchical_initialization_) {
-    std::uint64_t system_fingerprint = 0;
-    for (std::uint64_t fingerprint : celestial_jacobi_keplerian_fingerprints_) {
-      system_fingerprint = FingerprintCat2011(system_fingerprint, fingerprint);
-    }
-    LOG(INFO) << "System fingerprint is " << std::hex << system_fingerprint;
+  SolarSystem<Barycentric> solar_system(gravity_model_, initial_state_);
+
+  // If the system was constructed using keplerian elements, it may be the
+  // stock KSP system in which case it needs to be stabilized.
+  if (initial_state_.has_keplerian()) {
+    auto const hierarchical_system = solar_system.MakeHierarchicalSystem();
+    serialization::HierarchicalSystem message;
+    hierarchical_system->WriteToMessage(&message);
+    std::string const serialized_message = message.SerializeAsString();
+    uint64_t const system_fingerprint =
+        Fingerprint2011(serialized_message.c_str(), serialized_message.size());
```

Comment:
The serialized form of the `HierarchicalSystem` has both the bodies and their keplerian elements, so I believe that this covers everything.